### PR TITLE
add git compare api

### PR DESCRIPTION
**Pull request by a1012112796**
as title, I think it will be usefull for detect changed files in actions.

api format is similar with github, I hasn't found it in github's api document, but found an example link by the new bing:

https://api.github.com/repos/git/git/compare/v2.2.0-rc1...v2.2.0-rc2

### CODE REVIEW

File: routers/api/v1/repo/compare.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+				apiDiffFile.Status = "added"
+			} else if diffFile.Type == gitdiff.DiffFileDel {
+				apiDiffFile.Status = "deleted"
+			} else if diffFile.Type == gitdiff.DiffFileRename {
+				apiDiffFile.Status = "renamed"
+			} else if diffFile.Type == gitdiff.DiffFileChange {
+				apiDiffFile.Status = "modified"
+			} else if diffFile.Type == gitdiff.DiffFileCopy {
+				apiDiffFile.Status = "copied"
+			}
```
Comment (puni9869) [inline routers/api/v1/repo/compare.go]: what should be default case ? if applicable..

File: routers/api/v1/repo/compare.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+			if skipPatch {
+				result.Files = append(result.Files, apiDiffFile)
+				continue
+			}
+
+			for _, diffSec := range diffFile.Sections {
+				for _, diffLine := range diffSec.Lines {
+					apiDiffFile.Patch += diffLine.Content + "\n"
+				}
```
Comment (puni9869) [inline routers/api/v1/repo/compare.go]: Let's use stringbuilder here 
https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -1353,7 +1353,7 @@ func CompareAndPullRequestPost(ctx *context.Context) {
 		attachments []string
 	)
 
-	ci := ParseCompareInfo(ctx)
+	ci := parseCompareInfo(ctx)
```
Comment (puni9869) [inline routers/web/repo/pull.go:1356]: 💯

**Comment by puni9869** [comment]
Could you also add some unit test cases for this block 
```go
  if diffFile.Type == gitdiff.DiffFileAdd {
  apiDiffFile.Status = "added"
  } else if diffFile.Type == gitdiff.DiffFileDel {
  apiDiffFile.Status = "deleted"
  } else if diffFile.Type == gitdiff.DiffFileRename {
  apiDiffFile.Status = "renamed"
  } else if diffFile.Type == gitdiff.DiffFileChange {
  apiDiffFile.Status = "modified"
  } else if diffFile.Type == gitdiff.DiffFileCopy {
  apiDiffFile.Status = "copied"
  }
```

**Comment by puni9869** [comment]
I will review it again today.

**Comment by lchrennew** [comment]
Great, when will this api to be released

File: routers/api/v1/api.go [hunk 4]
```diff
@@ -1113,6 +1113,7 @@ func Routes() *web.Route {
 					m.Get("/tags/{sha}", repo.GetAnnotatedTag)
 					m.Get("/notes/{sha}", repo.GetNote)
 				}, context.ReferencesGitRepo(true), reqRepoReader(unit.TypeCode))
+				m.Get("/compare/*", context.ReferencesGitRepo(true), reqRepoReader(unit.TypeCode), repo.Compare)
```
Comment (puni9869) [inline routers/api/v1/api.go:1116]: Have a look on `web.go` `routers` file in `compare/*` Use the middleware if they are plugable to api route.

```
m.Combo("/compare/*", repo.MustBeNotEmpty, reqRepoCodeReader, repo.SetEditorconfigIfExists).
			Get(repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.CompareDiff).
			Post(reqSignIn, context.RepoMustNotBeArchived(), reqRepoPullsReader, repo.MustAllowPulls, web.Bind(forms.CreateIssueForm{}), repo.SetWhitespaceBehavior, repo.CompareAndPullRequestPost)
```

**Comment by lunny** [comment]
Replaced by #30349

**Pull request closed without merging**